<launch>
<arg name="uav" default="uav1"/>

<node name="global_network_node" pkg="dt_global_network_client" type="dt_global_network_client_node" output="screen">
    <param name="product_key" value="a1GcqS5WFte" />
    <param name="device_name" value="R_UAV_0" />
    <param name="device_secret" value="a516eead0501ed260f7caf87143b05d6"/>
    <param name="iot_pub_topic" value="/a1GcqS5WFte/R_UAV_0/user/R_UAV_0_PUB"/>
    <param name="iot_sub_topic" value="/a1GcqS5WFte/R_UAV_0/user/R_UAV_0_SUB"/>
    <param name="ros_sub_topic" value="/r_uav_0/msg_to_cloud"/>
    <param name="ros_pub_topic" value="/r_uav_0/msg_from_cloud"/>
    <param name="ros_iot_state_pub_topic" value="/r_uav_0/cloud_state"/>
    <param name="object_name" value="R_UAV_0"/>
  </node>

<node name="uav_msg_pack_center_$(arg uav)" pkg="message_pack" type="uav_message_pack_node" output="screen">
      <param name="SourceID" value="1"/>
      <param name="TargetID" value="201"/>
      <param name="IotMessagePubTopic" value="/r_uav_0/msg_to_cloud"/>
      <param name="GlobalPositionMessageSubTopic" value="/$(arg uav)/mavros/global_position/global"/>
      <param name="LocalPositionMessageSubTopic" value="/$(arg uav)/mavros/local_position/pose"/>
      <param name="LocalVelocityMessageSubTopic" value="/$(arg uav)/mavros/local_position/velocity_body"/>
      <param name="UavStateMessageSubTopic" value="/$(arg uav)/mavros/state"/>
      <param name="MessagePubFrequency" value="30"/>
      <param name="ObjectMotionModelMessageSubTopic" value="/$(arg uav)/model_learner/model_parameter"/>
      <param name="NetworkDelayTimeMessageSubTopic" value="/$(arg uav)/network_delay/recv_delay"/>
</node>

<node name="controller_network_wrapper_$(arg uav)" pkg="kf_mpc_controller4" type="controller_network_wrapper_node4" output="screen">
      <param name="network_message_sub_topic" value="/r_uav_0/msg_from_cloud"/>                          <!-- This topic data from network communication -->
      <param name="dt_object_position_pub_topic" value="/$(arg uav)/virtual/current_position"/>                    <!-- This topic data publish the dt object position-->
      <param name="virtual_guide_pub_topic" value="/$(arg uav)/virtual_guide/cmd"/>
      <param name="virtual_guide_control_pub_topic" value="/$(arg uav)/virtual_guide/type"/>
      <param name="dt_object_position_recv_hz" value="30.0"/><!-- This equal to virtual object position published frequency. -->
      <param name="is_unity" value="false"/>
      <param name="network_delay_pub_topic" value="/$(arg uav)/network_delay/recv_delay"/>
      <param name="network_delay_pub_hz" value="30.0"/>
      <param name="predict_window" value="8"/>                                                     <!--The window size for predicting further system state(predicted steps)-->
      <param name="predict_time_dt" value="0.033"/>                                                  <!--The predicted time step for least square-->
      <param name="target_position_pub_topic" value="/$(arg uav)/controller/target_position"/>
      <param name="model_learner_control_pub_topic" value="/model_learner/control"/>
</node>

<node name="mpc_controller_$(arg uav)" pkg="kf_mpc_controller4" type="mpc_controller_node4" output="screen">
      <param name="current_position_sub_topic" value="/$(arg uav)/mavros/local_position/pose"/>                          <!-- This topic data from sensor measure by GPS or UWB or other located system -->
      <param name="target_position_sub_topic" value="/$(arg uav)/controller/target_position"/>                           <!-- This tipic data from state predictor-->
      <param name="controller_output_pub_topic" value="/$(arg uav)/other_controller/target_velocity"/>              <!-- This topic data to px4 controller-->
      <param name="update_model_parameter_sub_topic" value="/$(arg uav)/model_learner/model_parameter"/>
      <param name="predict_time_dt_pub_topic" value="/$(arg uav)/controller/predict_time_dt"/>
      <param name="controller_running_hz" value="30"/>                                                        <!--The mpc controller running frequency -->
      <param name="state_dim" value="3"/>
      <param name="observe_dim" value="3"/>
      <param name="input_dim" value="3"/>
      <param name="control_window" value="5"/>
      <param name="predict_window" value="8"/>
      <param name="max_velocity" value = "2.0"/>
</node>

<node name="model_learner" pkg="kf_mpc_controller4" type="model_learner_node4" output="screen">
      <param name="local_position_sub_topic" value="/$(arg uav)/mavros/local_position/pose"/>                     <!-- This topic data from digital twin object -->
      <param name="target_velocity_sub_topic" value="/$(arg uav)/mavros/setpoint_velocity/cmd_vel_unstamped"/>                    <!-- This topic data from kalman filter-->
      <param name="update_model_parameter_pub_topic" value="/$(arg uav)/model_learner/model_parameter"/>                                                                                                        <!-- The gain of update measured error-->
      <param name="model_learner_info_pub_topic" value="/$(arg uav)/model_learner/information"/>
      <param name="model_learner_control_sub_topic" value="/model_learner/control"/>
      <param name="model_learner_run_hz" value="30.0"/>
      <param name="is_model_range" value="false"/>
      <param name="model_err_x" value = "0.00018"/>
      <param name="model_err_y" value = "0.00018"/>
      <param name="model_err_z" value = "0.00018"/>
      <param name="init_sample_time_x" value = "1"/>
      <param name="init_sample_time_y" value = "1"/>
      <param name="init_sample_time_z" value = "1"/>
</node>

<node name="fusion_controller" pkg="kf_mpc_controller4" type="fusion_controller_node2" output="screen">
      <param name="mpc_controller_output_sub_topic" value="/$(arg uav)/other_controller/target_velocity"/>
      <param name="virtual_guide_output_sub_topic" value="/$(arg uav)/virtual_guide/cmd"/>
      <param name="virtual_guide_control_sub_topic" value="/$(arg uav)/virtual_guide/type"/>
      <param name="virtual_guide_hz" value="30"/>
      <param name="target_velocity_pub_topic" value="/$(arg uav)/target/cmd"/>
      <param name="min_velocity" value="-2.0"/>
      <param name="max_velocity" value="2.0"/>
</node>

<node name="px4_velocity_control_wrapper_$(arg uav)" pkg="px4_simulation_wrapper" type="px4_velocity_control_wrapper" output="screen">
      <param name="px4_ref_vel_msg_pub_topic" value="/$(arg uav)/mavros/setpoint_velocity/cmd_vel_unstamped"/>        <!--The Target Velocity Message to PX4-->
      <param name="tgt_vel_msg_sub_topic" value="/$(arg uav)/target/cmd"/>                                            <!--The Target Velocity from Other Controller-->
      <param name="px4_state_msg_sub_topic" value="/$(arg uav)/mavros/state"/>                                        <!--The State of PX4-->
      <param name="px4_rc_msg_sub_topic" value="/$(arg uav)/mavros/rc/in"/>                                           <!--The RC Information from PX4-->
      <param name="px4_arming_client_topic" value="/$(arg uav)/mavros/cmd/arming"/>                                   <!--The Arm to PX4-->
      <param name="px4_set_mode_client_topic" value="/$(arg uav)/mavros/set_mode"/>                                   <!--The mode setting to PX4-->
      <param name="run_frequen_hz" value="50.0"/>                                                          <!--The Frequen of Target Velocity Message To PX4-->
      <param name="rc_start_ch" value="5"/>                                                                <!--The RC Channel of Start Vehicle-->
      <param name="rc_land_ch" value="6"/>                                                                 <!--The RC Channel of Land Vehicle-->
  </node>

<node name="rc_controller_$(arg uav)" pkg="rc_control" type="rc_control" output="screen">
      <param name="start_offboard_pub_topic" value="/$(arg uav)/mavros/rc/in"/>
      <param name="target_object_position_pub_topic" value = "/$(arg uav)/ref_uav/pose"/>
      <param name="start_offboard_ch" value="5"/>
      <param name="land_ch" value="6"/>
</node>

</launch>
